course3/lecture_en.md

```diff
@@ -0,0 +1,594 @@
+
+```moonbit expr
+fn fib1_mut(num: Int) -> Int64 {
+  // Declare a mutable variable with let mut
+  let mut map = @map.empty().insert(1, 1L).insert(2, 1L)
+  fn aux(num: Int) -> Int64 {
+    match map.lookup(num) {
+      Some(result) => result
+      None => {
+        let result_1 = aux(num - 1)
+        let result_2 = aux(num - 2)
+        // Update the binding with <variable> = <expression>
+        map = map.insert(num, result_1 + result_2) 
+        result_1 + result_2
+      }
+    }
+  }
+  aux(num)
+}
+```
+
+#### Bottom-Up Implementation
+
+In the bottom-up implementation, we typically start from the smallest subproblems, calculate and store the subsequent subproblems sequentially, until the original complex problem is solved.
+
+```moonbit expr
+fn fib2(num: Int) -> Int64 {
+  fn aux(n: Int, map: @map.Map[Int, Int64]) -> Int64 {
+    let result = get_or_else(map.lookup(n - 1), 1L) + 
+      get_or_else(map.lookup(n - 2), 1L)
+    if n == num { result }
+    else { aux(n + 1, map.insert(n, result)) }
+  }
+  let map = @map.empty().insert(0, 0L).insert(1, 1L)
+  aux(1, map)
+}
+```
+
+In particular, for the Fibonacci sequence, since we only need the results of the previous two subproblems when computing the current one, we can pass the results directly through the parameters, without the use of additional data structures.
+
+```moonbit expr
+fn fib2(num : Int) -> Int64 {
+  fn aux(n: Int, acc1: Int64, acc2: Int64) -> Int64 {
+    match n {
+      0 => acc1
+      _ => aux(n - 1, acc2, acc1 + acc2)
+    }
+  }
+  aux(num, 0L, 1L)
+}
+```
+
+The figure below visualizes the process of calculating $F_5$ and $F_4$ with bottom-up DP.
+![](../pics/fib_aux.drawio.png)
+
+## Summary
+
+In this module we learned
+- Basic data type: functions and their operations
+- Data structure: lists and pattern matching on lists
+- Algorithm: recursions and dynamic programming
+
+We also get an informal idea of ​​computational complexity.
+
+This course is intended to be an introductory level introduction. For a deeper understanding of how to use mathematical induction to prove the correctness of structured recursion, please refer to:
+- _**Software Foundations, Volume 1: Logical Foundations**_: Basics, Induction & Lists; or
+- _**Programming Language Foundations in Agda**_: Naturals, Induction & Relations
+
+If you want to learn more about algorithms and computational complexity, please refer to:
+- _**Algorithms** (4e)_: Chapter 1 - Fundamentals; or
+- _**Introduction to Algorithms** (4e)_: Chapter 3 - Characterizing Running Times; or
+- _**Introduction to Algorithms** (3e)_: Chapter 3 - Growth of Functions
+
+To learn more about dynamic programming, please refer to:
+- _**Introduction to Algorithms** (4e)_: Chapter 14 - Dynamic Programming; or
+- _**Introduction to Algorithms** (3e)_: Chapter 15 - Dynamic Programming
```

Comment:
I noticed that the readings listed in the Chinese slides and handout are different, and I have included a combination of the two here.